In [10]:
from google.cloud import bigquery
import pandas as pd

# Initialize BigQuery client
client = bigquery.Client(project="sql-sandbox-435415") 
#Project ID ensures that the code is reliable and portable, 
#as it explicitly tells BigQuery which project to use, allowing it to run consistently
#across different environments and for different users.

# Example query
query = """
SELECT
  *
FROM
  `bigquery-public-data.google_analytics_sample.ga_sessions_20170801`


LIMIT 10
"""

# Run the query and convert results to DataFrame
df = client.query(query).to_dataframe()
df


visitorId  visitNumber     visitId  visitStartTime      date  \
0       <NA>            1  1501591568      1501591568  20170801   
1       <NA>            2  1501589647      1501589647  20170801   
2       <NA>            1  1501616621      1501616621  20170801   
3       <NA>            1  1501601200      1501601200  20170801   
4       <NA>            1  1501615525      1501615525  20170801   
5       <NA>            1  1501610896      1501610896  20170801   
6       <NA>            1  1501589650      1501589650  20170801   
7       <NA>            1  1501573710      1501573710  20170801   
8       <NA>            1  1501644116      1501644116  20170801   
9       <NA>            1  1501574187      1501574187  20170801   

                                              totals  \
0  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
1  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
2  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
3  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
4  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
5  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
6  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
7  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
8  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   
9  {'visits': 1, 'hits': 1, 'pageviews': 1, 'time...   

                                       trafficSource  \
0  {'referralPath': None, 'campaign': '(not set)'...   
1  {'referralPath': '/analytics/web/', 'campaign'...   
2  {'referralPath': '/analytics/web/', 'campaign'...   
3  {'referralPath': '/analytics/web/', 'campaign'...   
4  {'referralPath': '/analytics/web/', 'campaign'...   
5  {'referralPath': None, 'campaign': '(not set)'...   
6  {'referralPath': '/analytics/web/', 'campaign'...   
7  {'referralPath': '/analytics/web/', 'campaign'...   
8  {'referralPath': None, 'campaign': '(not set)'...   
9  {'referralPath': None, 'campaign': '(not set)'...   

                                              device  \
0  {'browser': 'Chrome', 'browserVersion': 'not a...   
1  {'browser': 'Chrome', 'browserVersion': 'not a...   
2  {'browser': 'Chrome', 'browserVersion': 'not a...   
3  {'browser': 'Firefox', 'browserVersion': 'not ...   
4  {'browser': 'Chrome', 'browserVersion': 'not a...   
5  {'browser': 'Chrome', 'browserVersion': 'not a...   
6  {'browser': 'Chrome', 'browserVersion': 'not a...   
7  {'browser': 'Chrome', 'browserVersion': 'not a...   
8  {'browser': 'Chrome', 'browserVersion': 'not a...   
9  {'browser': 'Chrome', 'browserVersion': 'not a...   

                                          geoNetwork  \
0  {'continent': 'Europe', 'subContinent': 'South...   
1  {'continent': 'Asia', 'subContinent': 'Souther...   
2  {'continent': 'Europe', 'subContinent': 'North...   
3  {'continent': 'Americas', 'subContinent': 'Nor...   
4  {'continent': 'Americas', 'subContinent': 'Nor...   
5  {'continent': 'Americas', 'subContinent': 'Nor...   
6  {'continent': 'Asia', 'subContinent': 'Souther...   
7  {'continent': 'Asia', 'subContinent': 'Souther...   
8  {'continent': 'Americas', 'subContinent': 'Nor...   
9  {'continent': 'Americas', 'subContinent': 'Nor...   

                           customDimensions  \
0                                        []   
1           [{'index': 4, 'value': 'APAC'}]   
2           [{'index': 4, 'value': 'EMEA'}]   
3  [{'index': 4, 'value': 'North America'}]   
4  [{'index': 4, 'value': 'North America'}]   
5  [{'index': 4, 'value': 'North America'}]   
6           [{'index': 4, 'value': 'APAC'}]   
7           [{'index': 4, 'value': 'APAC'}]   
8  [{'index': 4, 'value': 'North America'}]   
9  [{'index': 4, 'value': 'North America'}]   

                                                hits        fullVisitorId  \
0  [{'hitNumber': 1, 'time': 0, 'hour': 5, 'minut...  3418334011779872055   
1  [{'hitNumber': 1, 'time': 0, 'hour': 5, 'minut...  2474397855041322408   
2  [{'hitNumber': 1, 'time': 0, 'hour': 12, 'min

This first query is used to connect to the Google Analytics database (in BigQuery) and preview the first 10 rows to understand what kind of data it contains.

In [11]:
query = """
SELECT
  count(*) as number_of_rows #count the number of rows
FROM
  `bigquery-public-data.google_analytics_sample.ga_sessions_20170801`
where
  totals.pageviews is not null 
"""

df = client.query(query).to_dataframe()
print (df)

   number_of_rows
0            2556


This query counts how many sessions occurred on August 1st, 2017, specifically where totals.pageviews is not null (i.e., sessions where at least one page was viewed).

There were 2,556 sessions with at least one pageview.

Running the same query without the WHERE condition gives the same result — this means all sessions on that day included at least one pageview.

In [13]:
query = """
SELECT
  distinct fullVisitorId as number_of_visitors #count the number of rows
FROM
  `bigquery-public-data.google_analytics_sample.ga_sessions_20170801`
"""

df = client.query(query).to_dataframe()
print (df)

       number_of_visitors
0     3418334011779872055
1     2474397855041322408
2     5870462820713110108
3     9397809171349480379
4     6089902943184578335
...                   ...
2288  5772921781547143127
2289  7550413639587874054
2290  3101662058536674321
2291  3754779580225027430
2292  5665035179702438805

[2293 rows x 1 columns]


Number of unique visitor in that day: 2293

Top Traffic Sources:
What to analyze:
Which sources (e.g., google, direct, referral) brought the most visitors?

Why:
To understand where users are coming from and which marketing channels perform best.

In [18]:
pd.set_option("display.max_rows", None)

query = """
SELECT DISTINCT trafficSource.source AS traffic_source
FROM `bigquery-public-data.google_analytics_sample.ga_sessions_20170801`
limit 35
  """

df = client.query(query).to_dataframe()
print (df)

              traffic_source
0                   (direct)
1       analytics.google.com
2         adwords.google.com
3                        ask
4   productforums.google.com
5                      baidu
6           sites.google.com
7            mail.google.com
8                  qiita.com
9      datastudio.google.com
10                google.com
11                       dfa
12                  Partners
13           blog.golang.org
14              sashihara.jp
15                 quora.com
16         groups.google.com
17              facebook.com
18               youtube.com
19               m.baidu.com
20       ph.search.yahoo.com
21            l.facebook.com
22        support.google.com
23            m.facebook.com
24             google.com.vn
25                     yahoo
26           lm.facebook.com
27                reddit.com
28             google.com.tw
29     int.search.tb.ask.com
30           docs.google.com
31                      bing
32               away.vk.com
33            

This query displays the traffic source where the customers arrived inside the eccomerce

In [21]:
query = """
SELECT
  trafficSource.medium AS traffic_channel,
  COUNT(*) AS sessions
FROM
  `bigquery-public-data.google_analytics_sample.ga_sessions_20170801`
GROUP BY
  traffic_channel
ORDER BY
  sessions DESC
LIMIT 10
"""

df = client.query(query).to_dataframe()
print (df)

  traffic_channel  sessions
0          (none)      2166
1        referral       313
2       affiliate        52
3             cpm        15
4         organic        10


- (none): Likely represents direct traffic (people typing the URL directly, using bookmarks, or missing referrer data).
- referral: Visitors who came from other websites via links.
- affiliate: Visitors referred by affiliate partners.
- cpm: Traffic from cost-per-thousand-impressions ad campaigns (likely display ads).
- organic: Search engine traffic (e.g., Google, Bing).
